shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -124,21 +160,47 @@ public void setMessageHandler(
       @Nullable TaskQueue taskQueue) {
     if (handler == null) {
       Log.v(TAG, "Removing handler for channel '" + channel + "'");
-      messageHandlers.remove(channel);
-    } else {
-      DartMessengerTaskQueue dartMessengerTaskQueue = null;
-      if (taskQueue != null) {
-        dartMessengerTaskQueue = createdTaskQueues.get(taskQueue);
-        if (dartMessengerTaskQueue == null) {
-          throw new IllegalArgumentException(
-              "Unrecognized TaskQueue, use BinaryMessenger to create your TaskQueue (ex makeBackgroundTaskQueue).");
-        }
+      synchronized (handlersLock) {
+        messageHandlers.remove(channel);
+      }
+      return;
+    }
+    DartMessengerTaskQueue dartMessengerTaskQueue = null;
+    if (taskQueue != null) {
+      dartMessengerTaskQueue = createdTaskQueues.get(taskQueue);
+      if (dartMessengerTaskQueue == null) {
+        throw new IllegalArgumentException(
+            "Unrecognized TaskQueue, use BinaryMessenger to create your TaskQueue (ex makeBackgroundTaskQueue).");
       }
-      Log.v(TAG, "Setting handler for channel '" + channel + "'");
+    }
+    Log.v(TAG, "Setting handler for channel '" + channel + "'");
+
+    LinkedList<BufferedMessageInfo> list;
+    synchronized (handlersLock) {
       messageHandlers.put(channel, new HandlerInfo(handler, dartMessengerTaskQueue));
+      if (!bufferedMessages.containsKey(channel)) {
+        return;
+      }
+      list = (LinkedList) bufferedMessages.get(channel);
+      bufferedMessages.remove(channel);
+    }
+    for (BufferedMessageInfo info : list) {
```

Comment:
This isn't thread-safe.  The ui thread could be adding to the linked list while the platform thread is iterating over it.

Reply:
That case would be handled by the normal flow. Since L180 already populated messageHandlers, there should be no issue AFAIT

Reply:
The problem isn't `messageHandlers`, the LinkedList data structure can be simultaneously written to and read from 2 different threads at the same time.  Here it is being read on L187 unguarded and it can be written to on L305.  Reading and writing to a non-threadsafe data structure simultaneously can result in unpredictable results.

Reply:
`synchronized(lock)` is handling the locking. Is there an issue with this lock?

Reply:
Yes, that's what I'm telling you =)   Specifically there is no lock while you are reading from the LinkedList, but there is one when you are writing to it.  When you read the List from bufferedMessages it is passed by reference, not by value, it's the same datastructure being written to and read from, only the write has a lock.

Reply:
My understanding of this problem is:
1. Once we exited the synchronization block above, `messageHandlers` has been set. At this point, a thread trying to get the lock in the second synchronization block runs.
2. The second synchronization block realizes that the `messageHandler` is in the map, therefore `messageDeferred` returns `false`, and no write to the buffer list occurs. 

As a result, there's no case in which this loop is running and a write occurs in the middle.

Reply:
Right, because of the removal and acquisition of the list are guarded, this usage is correct.  What I've said above is true but isn't a problem in practice.  There is no guarantee that someone isn't accessing the LinkedList while we are reading from it, but we don't do it.  I'm trying to think if there is any change I recommend, a comment or something.

If `bufferedMessages` was a ConcurrentMap and the LinkedList implementation was persistent you could guarantee correctness (unfortunately there isn't a persistent linked list implementation in the sdk).  Did you not want to use ConcurrentMap because you didn't want the overhead of locking twice?  That's probably the right call.

This is fine with me, I can't think of a comment or easy change that makes this better given what the sdk provides.  Thanks for walking through it.

Reply:
right. Do you know if `ConcurrentMap` is actually preventing races? This part from the docs is confusing:

> Retrieval operations (including get) generally do not block, so may overlap with update operations (including put and remove)

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html

Reply:
Yea, it's thread-safe.  That just means it is implemented in such a way that you won't have the overhead of a lock when doing read operations which is cool, that's what we'd want.  Using a lock, reading and writing would be serialized.  It also means that technically a read and a write at the same exact time might not strictly happen in order.  That's why it has methods that read and write the value atomically (like merge).